File: src/llm_cal/model_source/modelscope.py
"""ModelScope source — REST-only via httpx.

Decision: Option B from ADR-001. We don't need the official `modelscope` SDK
because llm-infer-cal only requires three things:
  1. List repo files + sizes (one API call)
  2. Fetch config.json (one API call)
  3. Range-GET a safetensors header (handled by safetensors_reader)

The SDK pulls heavy ML deps by default (torch / tf for some install paths).
REST keeps the install footprint flat, mirrors the existing httpx hot path,
and gives us identical exception semantics across HF + MS.

Endpoints (verified against modelscope.cn public docs, 2026-04):
  * GET /api/v1/models/{owner}/{name}                       — model meta
  * GET /api/v1/models/{owner}/{name}/repo/files?Recursive=true
                                                            — file tree + sizes
  * GET /api/v1/models/{owner}/{name}/repo?FilePath=...&Revision=...
                                                            — raw file content

ModelScope wraps every response in a {Code, Message, Data, Success} envelope.
Field casing is PascalCase. We parse defensively — fields may evolve.
"""

from __future__ import annotations

import json
from typing import Any

import httpx

from llm_cal.model_source.auth import (
    get_modelscope_token,
    modelscope_auth_error_message,
)
from llm_cal.model_source.base import (
    AuthRequiredError,
    ModelArtifact,
    ModelNotFoundError,
    ModelSource,
    SiblingFile,
    SourceUnavailableError,
)

DEFAULT_ENDPOINT = "https://www.modelscope.cn"
DEFAULT_REVISION = "master"

_INFO_PATH = "/api/v1/models/{model_id}"
_FILES_PATH = "/api/v1/models/{model_id}/repo/files"
_RAW_PATH = "/api/v1/models/{model_id}/repo"


class ModelScopeSource(ModelSource):
    name = "modelscope"

    def __init__(
        self,
        endpoint: str | None = None,
        timeout_s: float = 30.0,
        revision: str = DEFAULT_REVISION,
    ) -> None:
        self._endpoint = (endpoint or DEFAULT_ENDPOINT).rstrip("/")
        self._timeout_s = timeout_s
        self._revision = revision

    def fetch(self, model_id: str) -> ModelArtifact:
        token = get_modelscope_token()
        headers = self._auth_headers(token)

        # Step 1: model info — gives us LatestSha (commit pin) when available.
        # We tolerate missing info; fall back to revision="master" so that the
        # file list + config calls still work.
        commit_sha = self._fetch_commit_sha(model_id, headers)

        # Step 2: file tree with sizes. ONE call, recursive, includes sub-folders.
        siblings = self._list_files(model_id, commit_sha or self._revision, headers)

        # Step 3: config.json. Pin to the commit sha when we have it so two
        # back-to-back calls don't race against a repo update.
        config = self._fetch_config(model_id, commit_sha or self._revision, headers)

        return ModelArtifact(
            source=self.name,
            model_id=model_id,
            commit_sha=commit_sha,
            config=config,
            siblings=siblings,
        )

    # ------------------------------------------------------------------ helpers

    def _auth_headers(self, token: str | None) -> dict[str, str]:
        return {"Authorization": f"Bearer {token}"} if token else {}

    def _fetch_commit_sha(self, model_id: str, headers: dict[str, str]) -> str | None:
        url = f"{self._endpoint}{_INFO_PATH.format(model_id=model_id)}"
        try:
            resp = httpx.get(
                url, headers=headers, timeout=self._timeout_s, follow_redirects=True
            )
        except (httpx.TimeoutException, httpx.ConnectError, httpx.HTTPError):
            # Soft fail — commit sha is best-effort. Caller will use "master".
            return None

        if resp.status_code != 200:
            return None
        try:
            payload = resp.json()
        except json.JSONDecodeError:
            return None

        data = payload.get("Data") if isinstance(payload, dict) else None
        if not isinstance(data, dict):
            return None
        # Field name has bounced between LatestSha / latest_sha / Revision in
        # historical docs; check several.
        for key in ("LatestSha", "latest_sha", "Revision", "Sha"):
            v = data.get(key)
            if isinstance(v, str) and v:
                return v
        return None

    def _list_files(
        self, model_id: str, revision: str, headers: dict[str, str]
    ) -> tuple[SiblingFile, ...]:
        url = f"{self._endpoint}{_FILES_PATH.format(model_id=model_id)}"
        params = {"Recursive": "true", "Revision": revision}
        try:
            resp = httpx.get(
                url,
                headers=headers,
                params=params,
                timeout=self._timeout_s,
                follow_redirects=True,
            )
        except (httpx.TimeoutException, httpx.ConnectError) as e:
            raise SourceUnavailableError(f"ModelScope file list failed: {e}") from e

        self._raise_for_status(resp, model_id, what="file list")

        try:
            payload = resp.json()
        except json.JSONDecodeError as e:
            raise SourceUnavailableError(
                f"ModelScope file list returned non-JSON: {e}"
            ) from e

        files = _extract_files(payload)
        if files is None:
            raise SourceUnavailableError(
                "ModelScope file list payload had unexpected shape — "
                "neither Data.Files nor Data is a list."
            )
        return tuple(
            SiblingFile(filename=f["Path"], size=f.get("Size"))
            for f in files
            if isinstance(f, dict) and isinstance(f.get("Path"), str)
            # Only include blobs (not directories). Type=tree means folder.
            and f.get("Type", "blob") != "tree"
        )

    def _fetch_config(
        self, model_id: str, revision: str, headers: dict[str, str]
    ) -> dict[str, Any]:
        url = f"{self._endpoint}{_RAW_PATH.format(model_id=model_id)}"
        params = {"FilePath": "config.json", "Revision": revision}
        try:
            resp = httpx.get(
                url,
                headers=headers,
                params=params,
                timeout=self._timeout_s,
                follow_redirects=True,
            )
        except (httpx.TimeoutException, httpx.ConnectError) as e:
            raise SourceUnavailableError(f"config.json fetch failed: {e}") from e

        self._raise_for_status(resp, model_id, what="config.json")

        try:
            parsed: Any = json.loads(resp.text)
        except json.JSONDecodeError as e:
            raise SourceUnavailableError(
                f"config.json is not valid JSON (line {e.lineno} col {e.colno}): {e.msg}"
            ) from e
        if not isinstance(parsed, dict):
            raise SourceUnavailableError(
                "config.json did not parse to a JSON object."
            )
        return parsed

    def _raise_for_status(
        self, resp: httpx.Response, model_id: str, what: str
    ) -> None:
        if resp.status_code == 200:
            return
        if resp.status_code == 404:
            raise ModelNotFoundError(
                f"Model '{model_id}' not found on ModelScope ({what})."
            )
        if resp.status_code in (401, 403):
            raise AuthRequiredError(modelscope_auth_error_message(model_id))
        if resp.status_code == 429:
            retry = resp.headers.get("Retry-After", "unknown")
            raise SourceUnavailableError(
                f"ModelScope rate limit (429). Retry-After: {retry}s. "
                "Setting MODELSCOPE_API_TOKEN increases your quota."
            )
        raise SourceUnavailableError(
            f"ModelScope {what} returned HTTP {resp.status_code}"
        )


def _extract_files(payload: Any) -> list[Any] | None:
    """Pull the file list out of the wrapped ModelScope envelope.

    Tolerates two known shapes:
      A) {Data: {Files: [...]}}      — most common
      B) {Data: [...]}               — older / list-only endpoints
    """
    if not isinstance(payload, dict):
        return None
    data = payload.get("Data")
    if isinstance(data, dict):
        files = data.get("Files")
        if isinstance(files, list):
            return files
    if isinstance(data, list):
        return data
    return None

File: crates/llm-infer-cal-core/src/model_source/modelscope.rs
use std::collections::HashMap;
use std::time::Duration;

use serde_json::Value;

use crate::model_source::auth::{get_modelscope_token, modelscope_auth_error_message};
use crate::model_source::base::{
    AuthRequiredError, ModelArtifact, ModelNotFoundError, ModelSource, ModelSourceError,
    SiblingFile, SourceUnavailableError,
};

pub const DEFAULT_ENDPOINT: &str = "https://www.modelscope.cn";
pub const DEFAULT_REVISION: &str = "master";

const INFO_PATH: &str = "/api/v1/models/{model_id}";
const FILES_PATH: &str = "/api/v1/models/{model_id}/repo/files";
const RAW_PATH: &str = "/api/v1/models/{model_id}/repo";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HttpResponse {
    pub status: u16,
    pub headers: HashMap<String, String>,
    pub body: Vec<u8>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HttpError(pub String);

pub trait HttpClient {
    fn get(
        &mut self,
        url: &str,
        headers: &HashMap<String, String>,
        params: &[(String, String)],
        timeout_s: f64,
    ) -> Result<HttpResponse, HttpError>;
}

#[derive(Default)]
pub struct ReqwestHttpClient;

impl HttpClient for ReqwestHttpClient {
    fn get(
        &mut self,
        url: &str,
        headers: &HashMap<String, String>,
        params: &[(String, String)],
        timeout_s: f64,
    ) -> Result<HttpResponse, HttpError> {
        let client = reqwest::blocking::Client::builder()
            .timeout(Duration::from_secs_f64(timeout_s))
            .build()
            .map_err(|error| HttpError(error.to_string()))?;
        let mut request = client.get(url).query(params);
        for (key, value) in headers {
            request = request.header(key, value);
        }
        let response = request
            .send()
            .map_err(|error| HttpError(error.to_string()))?;
        let status = response.status().as_u16();
        let mut response_headers = HashMap::new();
        for (key, value) in response.headers() {
            response_headers.insert(
                key.as_str().to_string(),
                value.to_str().unwrap_or_default().to_string(),
            );
        }
        let body = response
            .bytes()
            .map_err(|error| HttpError(error.to_string()))?
            .to_vec();
        Ok(HttpResponse {
            status,
            headers: response_headers,
            body,
        })
    }
}

#[derive(Clone, Debug)]
pub struct ModelScopeSource {
    endpoint: String,
    timeout_s: f64,
    revision: String,
}

impl Default for ModelScopeSource {
    fn default() -> Self {
        Self::new(None, 30.0, DEFAULT_REVISION)
    }
}

impl ModelScopeSource {
    pub fn new(endpoint: Option<&str>, timeout_s: f64, revision: &str) -> Self {
        Self {
            endpoint: endpoint
                .unwrap_or(DEFAULT_ENDPOINT)
                .trim_end_matches('/')
                .to_string(),
            timeout_s,
            revision: revision.to_string(),
        }
    }

    pub fn fetch(&self, model_id: &str) -> Result<ModelArtifact, ModelSourceError> {
        let token = get_modelscope_token();
        let mut client = ReqwestHttpClient;
        self.fetch_with_client(model_id, &mut client, token.as_deref())
    }

    pub fn fetch_with_client(
        &self,
        model_id: &str,
        client: &mut dyn HttpClient,
        token: Option<&str>,
    ) -> Result<ModelArtifact, ModelSourceError> {
        let headers = auth_headers(token);
        let commit_sha = self.fetch_commit_sha(model_id, client, &headers);
        let revision = commit_sha.as_deref().unwrap_or(&self.revision);

        let siblings = self.list_files(model_id, revision, client, &headers)?;
        let config = self.fetch_config(model_id, revision, client, &headers)?;

        Ok(ModelArtifact {
            source: "modelscope".to_string(),
            model_id: model_id.to_string(),
            commit_sha,
            config,
            siblings,
        })
    }

    fn fetch_commit_sha(
        &self,
        model_id: &str,
        client: &mut dyn HttpClient,
        headers: &HashMap<String, String>,
    ) -> Option<String> {
        let url = self.url(INFO_PATH, model_id);
        let response = client.get(&url, headers, &[], self.timeout_s).ok()?;
        if response.status != 200 {
            return None;
        }
        let payload: Value = serde_json::from_slice(&response.body).ok()?;
        let data = payload.get("Data")?.as_object()?;
        for key in ["LatestSha", "latest_sha", "Revision", "Sha"] {
            if let Some(value) = data
                .get(key)
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
            {
                return Some(value.to_string());
            }
        }
        None
    }

    fn list_files(
        &self,
        model_id: &str,
        revision: &str,
        client: &mut dyn HttpClient,
        headers: &HashMap<String, String>,
    ) -> Result<Vec<SiblingFile>, ModelSourceError> {
        let url = self.url(FILES_PATH, model_id);
        let params = vec![
            ("Recursive".to_string(), "true".to_string()),
            ("Revision".to_string(), revision.to_string()),
        ];
        let response = client
            .get(&url, headers, &params, self.timeout_s)
            .map_err(|error| {
                SourceUnavailableError(format!("ModelScope file list failed: {}", error.0))
            })?;

        self.raise_for_status(&response, model_id, "file list")?;
        let payload: Value = serde_json::from_slice(&response.body).map_err(|error| {
            SourceUnavailableError(format!("ModelScope file list returned non-JSON: {error}"))
        })?;
        let files = extract_files(&payload).ok_or_else(|| {
            SourceUnavailableError(
                "ModelScope file list payload had unexpected shape — neither Data.Files nor Data is a list."
                    .to_string(),
            )
        })?;

        Ok(files
            .into_iter()
            .filter_map(|file| {
                let object = file.as_object()?;
                if object.get("Type").and_then(Value::as_str).unwrap_or("blob") == "tree" {
                    return None;
                }
                let filename = object.get("Path").and_then(Value::as_str)?.to_string();
                let size = object.get("Size").and_then(value_to_u64);
                Some(SiblingFile { filename, size })
            })
            .collect())
    }

    fn fetch_config(
        &self,
        model_id: &str,
        revision: &str,
        client: &mut dyn HttpClient,
        headers: &HashMap<String, String>,
    ) -> Result<Value, ModelSourceError> {
        let url = self.url(RAW_PATH, model_id);
        let params = vec![
            ("FilePath".to_string(), "config.json".to_string()),
            ("Revision".to_string(), revision.to_string()),
        ];
        let response = client
            .get(&url, headers, &params, self.timeout_s)
            .map_err(|error| {
                SourceUnavailableError(format!("config.json fetch failed: {}", error.0))
            })?;

        self.raise_for_status(&response, model_id, "config.json")?;
        let parsed: Value = serde_json::from_slice(&response.body).map_err(|error| {
            SourceUnavailableError(format!("config.json is not valid JSON: {error}"))
        })?;
        if !parsed.is_object() {
            return Err(SourceUnavailableError(
                "config.json did not parse to a JSON object.".to_string(),
            )
            .into());
        }
        Ok(parsed)
    }

    fn raise_for_status(
        &self,
        response: &HttpResponse,
        model_id: &str,
        what: &str,
    ) -> Result<(), ModelSourceError> {
        match response.status {
            200 => Ok(()),
            404 => Err(ModelNotFoundError(format!(
                "Model '{model_id}' not found on ModelScope ({what})."
            ))
            .into()),
            401 | 403 => Err(AuthRequiredError(modelscope_auth_error_message(model_id)).into()),
            429 => {
                let retry = response
                    .headers
                    .get("Retry-After")
                    .or_else(|| response.headers.get("retry-after"))
                    .map(String::as_str)
                    .unwrap_or("unknown");
                Err(SourceUnavailableError(format!(
                    "ModelScope rate limit (429). Retry-After: {retry}s. Setting MODELSCOPE_API_TOKEN increases your quota."
                ))
                .into())
            }
            status => Err(SourceUnavailableError(format!(
                "ModelScope {what} returned HTTP {status}"
            ))
            .into()),
        }
    }

    fn url(&self, path: &str, model_id: &str) -> String {
        format!("{}{}", self.endpoint, path.replace("{model_id}", model_id))
    }
}

impl ModelSource for ModelScopeSource {
    fn name(&self) -> &str {
        "modelscope"
    }

    fn fetch(&self, model_id: &str) -> Result<ModelArtifact, ModelSourceError> {
        ModelScopeSource::fetch(self, model_id)
    }
}

pub fn extract_files(payload: &Value) -> Option<Vec<Value>> {
    let data = payload.as_object()?.get("Data")?;
    if let Some(data) = data.as_object() {
        if let Some(files) = data.get("Files").and_then(Value::as_array) {
            return Some(files.clone());
        }
    }
    data.as_array().cloned()
}

fn auth_headers(token: Option<&str>) -> HashMap<String, String> {
    token
        .filter(|token| !token.is_empty())
        .map(|token| HashMap::from([("Authorization".to_string(), format!("Bearer {token}"))]))
        .unwrap_or_default()
}

fn value_to_u64(value: &Value) -> Option<u64> {
    match value {
        Value::Number(number) => number.as_u64(),
        Value::String(text) => text.parse().ok(),
        _ => None,
    }
}
